File: src-tauri/src/messenger/js/peek.js
// Modal "peek" frame for the link-preview panel (see commands.rs::open_peek).
// The preview content itself is a NATIVE child webview that always paints above
// this page, so this only draws the dim backdrop and the header bar that sit
// AROUND it: the child webview rect is inset below the header.
//
// This page is ALSO the source of truth for the child webview's geometry. Rust
// can't place the native child in the page's CSS pixels directly (its
// scale_factor / inner_size are logical px, which diverge from CSS px under
// HiDPI scaled display modes -> FB spills past the frame). So we measure the
// hole here in CSS px and report it via `swissknife-link://peek?x&y&w&h`; Rust
// sizes the child to exactly that. MARGIN/HEADER must match messenger.css.
// Close routes through the shared swissknife-link:// scheme, the same channel
// the title bar uses, so the remote page still gets no IPC access.
(function () {
  var ID = "__sk_peek";
  var MARGIN = 24;
  var HEADER = 40;
  var active = false;
  function close() {
    window.location.href = "swissknife-link://window?action=peek-close";
  }
  // Report the modal hole (in CSS px) so Rust can size the child webview to it.
  function reportRect() {
    if (!active) return;
    var w = window.innerWidth;
    var h = window.innerHeight;
    var y = MARGIN + HEADER;
    var cw = Math.max(120, w - 2 * MARGIN);
    var ch = Math.max(120, h - y - MARGIN);
    window.location.href =
      "swissknife-link://peek?x=" + MARGIN + "&y=" + y + "&w=" + cw + "&h=" + ch;
  }
  var pending = false;
  function scheduleReport() {
    if (pending || !active) return;
    pending = true;
    requestAnimationFrame(function () {
      pending = false;
      reportRect();
    });
  }
  function build() {
    var existing = document.getElementById(ID);
    if (existing || !document.body) return existing;
    var back = document.createElement("div");
    back.id = ID;
    // Click on the dim margin (anywhere the child webview doesn't cover) closes.
    back.addEventListener("click", close);
    var bar = document.createElement("div");
    bar.id = "__sk_peek_bar";
    bar.addEventListener("click", function (e) { e.stopPropagation(); });
    var label = document.createElement("span");
    label.id = "__sk_peek_url";
    var x = document.createElement("button");
    x.type = "button";
    x.id = "__sk_peek_close";
    x.setAttribute("aria-label", "Close preview");
    x.textContent = "×";
    x.addEventListener("click", close);
    bar.appendChild(label);
    bar.appendChild(x);
    back.appendChild(bar);
    document.body.appendChild(back);
    return back;
  }
  window.__skPeekShow = function (url) {
    active = true;
    window.__skPeekActive = true;
    var back = build();
    if (!back) return;
    var label = document.getElementById("__sk_peek_url");
    if (label) label.textContent = url || "";
    back.classList.add("sk-on");
    reportRect();
  };
  window.__skPeekHide = function () {
    active = false;
    window.__skPeekActive = false;
    var back = document.getElementById(ID);
    if (back) back.classList.remove("sk-on");
  };
  // Keep the child webview matched to the window as it resizes.
  window.addEventListener("resize", scheduleReport);
  // Esc closes the preview instead of collapsing the window (bubble.js defers to
  // us via __skPeekActive). Capture phase so we win over Facebook's handlers.
  // This only fires while focus is in THIS page; once the user clicks into the
  // panel, focus moves to the native child webview, which has its own Esc
  // handler (see commands.rs::PEEK_ESC_JS).
  document.addEventListener(
    "keydown",
    function (e) {
      if (active && e.key === "Escape") {
        e.preventDefault();
        e.stopImmediatePropagation();
        close();
      }
    },
    true
  );
  // Facebook is an SPA that can wipe the DOM; re-add the frame while a preview is
  // open so it doesn't vanish under the (still-native) child webview.
  setInterval(function () {
    if (!active) return;
    var back = build();
    if (back) back.classList.add("sk-on");
  }, 1500);
})();

File: src-tauri/src/messenger/peek.rs
//! Link-preview panel: a native child webview overlaid on the Messenger window
//! as an in-window modal. `peek.js` draws the frame; this manages the child.

use tauri::{AppHandle, LogicalPosition, LogicalSize, Manager, Url, WebviewBuilder, WebviewUrl};

use super::window::DESKTOP_UA;
use crate::messenger::{MESSENGER_LABEL, PEEK_LABEL, ROUTE_SCHEME};

/// Injected into the preview child webview so Esc closes it even when the panel
/// (not the Messenger page) has keyboard focus. Routes through the shared scheme,
/// caught by the child's own `on_navigation` -> `close_peek`.
const PEEK_ESC_JS: &str = "document.addEventListener('keydown',function(e){if(e.key==='Escape'){window.location.href='swissknife-link://window?action=peek-close';}},true);";

/// Inset of the link-preview modal from the Messenger window edges, and the
/// height of its header bar. The preview child webview is placed in the rect
/// below the header; `peek.js` / `messenger.css` draw the frame around it, so
/// these MUST stay in sync with the geometry in `messenger.css`.
const PEEK_MARGIN: f64 = 24.0;
const PEEK_HEADER: f64 = 40.0;

/// Open a URL in the link-preview panel: a native child webview overlaid on the
/// Messenger window (an in-window modal, not a separate OS window). A native
/// webview loads sites that refuse iframing (Facebook, banks, ...), which an
/// in-page iframe modal can't. Reuses the open preview if there is one (navigate
/// and resize), else attaches a new child. `peek.js` draws the modal frame
/// backdrop and header around it in the Messenger page.
pub(super) fn open_peek(app: &AppHandle, url: &str) {
    let target: Url = match url.parse() {
        Ok(u) => u,
        Err(_) => return,
    };
    let Some(mwin) = app.get_webview_window(MESSENGER_LABEL) else {
        return;
    };
    // Seed geometry only: peek.js reports the exact CSS-px rect once shown and
    // resizes the child to it, so the child never renders at this size.
    let pos = LogicalPosition::new(PEEK_MARGIN, PEEK_MARGIN + PEEK_HEADER);
    let size = LogicalSize::new(400.0, 400.0);
    // Reuse an already-open preview: navigate to the new target.
    if let Some(wv) = app.get_webview(PEEK_LABEL) {
        let _ = wv.navigate(target);
        let _ = wv.set_focus();
    } else {
        // Attach a fresh child webview to the Messenger window (multiwebview; needs
        // the tauri `unstable` feature). `peek_bounds` is only a seed size; peek.js
        // reports the exact CSS-px rect once shown (see the `peek` host above).
        // Only our close scheme is intercepted; every other link navigates within
        // the preview.
        let Some(win) = app.get_window(MESSENGER_LABEL) else {
            return;
        };
        let esc_handle = app.clone();
        let child = win.add_child(
            WebviewBuilder::new(PEEK_LABEL, WebviewUrl::External(target))
                .user_agent(DESKTOP_UA)
                .initialization_script(PEEK_ESC_JS)
                .on_navigation(move |u| {
                    if u.scheme() == ROUTE_SCHEME {
                        close_peek(&esc_handle);
                        return false;
                    }
                    true
                }),
            pos,
            size,
        );
        // Clip FB to the child's frame so it can't paint past the modal. All four
        // corners rounded (a card under the header); masksToBounds does the clip.
        #[cfg(target_os = "macos")]
        if let Ok(child) = child {
            let _ = child.with_webview(|pw| {
                super::window::mask_webview(pw.inner() as *mut objc2::runtime::AnyObject, 10.0, 0);
            });
        }
        #[cfg(not(target_os = "macos"))]
        let _ = child;
    }
    // Draw the modal frame (dim backdrop + header) in the Messenger page and let
    // it report the exact child rect. Done after the child exists so the report
    // lands on a live webview.
    let _ = mwin.eval(format!(
        "window.__skPeekShow&&window.__skPeekShow({})",
        serde_json::to_string(url).unwrap_or_default()
    ));
}

/// Close the link-preview panel: destroy the child webview and hide its modal
/// frame. Reached from the frame's close button / backdrop (`peek-close`) and
/// when the window collapses to a bubble.
pub fn close_peek(app: &AppHandle) {
    if let Some(wv) = app.get_webview(PEEK_LABEL) {
        let _ = wv.close();
    }
    if let Some(mwin) = app.get_webview_window(MESSENGER_LABEL) {
        let _ = mwin.eval("window.__skPeekHide&&window.__skPeekHide()");
    }
}
